### PR TITLE
Reusable dispatcher

### LINKED ISSUE #401: raise RuntimeError('Connection Pool already initialized') after restarting update instances
### Steps to reproduce
1.  Using multiple telegram bot update instance in gevent based server 
2. Start them with updater.start_polling()
2. Stop all updater.stop()
3. Restart them with  updater.start_polling()
 
### Expected behaviour
Make possible to run mutliple updater instance with either shared urlib3 pools or seperate pools
 
### Actual behaviour
Failed to allocate new dispatcher workers due global constraint in urllib3 connection pool

### Configuration
Linux

**Version of Python, python-telegram-bot & dependencies:**

```
urllib3 1.10 has no metadata
python-telegram-bot 5.0.0 has no metadata
```

### Logs
```
  File "/opt/devcores/env/local/lib/python2.7/site-packages/telegram/ext/updater.py", line 85, in __init__
    exception_event=self.__exception_event)
  File "/opt/devcores/env/local/lib/python2.7/site-packages/telegram/ext/dispatcher.py", line 124, in __init__
    raise RuntimeError('Connection Pool already initialized')
```

### PR DESCRIPTION
**Pull request by tsnoam**
Fixes #401

### CODE REVIEW

File: telegram/ext/updater.py [hunk 1]
```diff
@@ -74,7 +79,13 @@ def __init__(self, token=None, base_url=None, workers=4, bot=None):
         if bot is not None:
             self.bot = bot
         else:
-            self.bot = Bot(token, base_url)
+            # we need a connection pool the size of:
+            # * for each of the workers
+            # * 1 for Dispatcher
+            # * 1 for polling Updater (even if webhook is used, we can spare a connection)
+            # * 1 for JobQueue
+            self._request = Request(con_pool_size=workers + 3)
```
Comment (jh0ker) [inline telegram/ext/updater.py]: This would be a good occasion to add the connection for the main thread (so +4)

File: tests/test_updater.py [hunk 2]
```diff
@@ -710,6 +753,8 @@ def __init__(self,
         self.bootstrap_attempts = 0
         self.bootstrap_err = bootstrap_err
         self.edited = edited
+        self.request = MockyMock()
```
Comment (jh0ker) [inline tests/test_updater.py]: Is there a reason we can't use `None` here?

**Comment by jh0ker** [comment]
(Edited and extended) copy from the dev chat:
Honestly, big props that you actually managed to make the singleton thread pool used for the `run_async` decorator optional.

Maybe we should remove the singleton completely in the next major version and go full class based. The downside would be that the dispatcher had to be declared globally then,  which is not nice.

Still, here is a suggestion.

Right now, we have `Dispatcher.__async_queue` and `Dispatcher.__async_threads` with this behaviour:

- In case of singleton-mode, `_ASYNC_QUEUE` and `_ASYNC_THREADS`, which are declared on module level, are assigned to the (only one) `Dispatcher` instance
- In case of multiple instances (with `no_singleton` set to `True`), a new `Queue` and `set` instance are created for each instance

I'm wondering if the following behaviour would be preferrable, especially for future development:
- Remove the module-level declarations and implement a "classic" singleton pattern by having a module-level variable called `INSTANCE` and a `threading.Event` instance called `SINGLETON_MODE` that is `set` by default

- When the `Dispatcher` is instantiated first, set the `INSTANCE` to `self`
- If a second `Dispatcher` is instantiated, `clear` the `SINGLETON_MODE` event
- In `run_async`, check if the `SINGLETON_MODE` event is set. If it is, proceed as normal and use the `INSTANCE` variable to enqueue the function. If not, raise an exception.

In my opinion, that is more explicit and obvious than the current implementation and I think will be more maintainable. 
The current implementation also allows to use the `run_async` decorator even if multiple `Dispatcher` instances are present (correct me if I'm wrong here). To me, it also smells of hotfix a bit too much ;)

Let me know what you think, and I'm sure you have some great ideas on how to improve my proposal in case you agree with the general idea.

File: tests/test_updater.py (hunk 2 above)
Comment (tsnoam) [inline tests/test_updater.py]: @jh0ker 
This is actually leftovers from a stage in the development. I just removed it now...

File: telegram/ext/updater.py (hunk 1 above)
Comment (tsnoam) [inline telegram/ext/updater.py]: @jh0ker 
I don't think we need it... but it probably doesn't hurt to add another connection.

**Comment by tsnoam** [comment]
@jh0ker 
I like your approach for this. I'll put it into code and lets see what we'll get :)

**Comment by jh0ker** [comment]
I love it! Really good take on the idea, this is exactly what I had in mind, just better :laughing:

**Comment by jh0ker** [comment]
LGTM :+1:

**Pull request merged**